### ISSUE TITLE
[Bug]: Unable to resolve Payment Module Service in my Loader Function

### ISSUE THREAD

**Issue by okechukwu0127** [medusajs/medusa#10776]
### Package.json file

```JSON
{
  "name": "medusa-starter-default",
  "version": "0.0.1",
  "description": "A starter for Medusa projects.",
  "author": "Medusa (https://medusajs.com)",
  "license": "MIT",
  "keywords": [
    "sqlite",
    "postgres",
    "typescript",
    "ecommerce",
    "headless",
    "medusa"
  ],
  "scripts": {
    "build": "medusa build",
    "seed": "medusa exec ./src/scripts/seed.ts",
    "start": "medusa start",
    "dev": "medusa develop",
    "test:integration:http": "TEST_TYPE=integration:http NODE_OPTIONS=--experimental-vm-modules jest --silent=false --runInBand --forceExit",
    "test:integration:modules": "TEST_TYPE=integration:modules NODE_OPTIONS=--experimental-vm-modules jest --silent --runInBand --forceExit",
    "test:unit": "TEST_TYPE=unit NODE_OPTIONS=--experimental-vm-modules jest --silent --runInBand --forceExit"
  },
  "dependencies": {
    "@medusajs/admin-sdk": "latest",
    "@medusajs/cli": "latest",
    "@medusajs/framework": "latest",
    "@medusajs/medusa": "latest",
    "@mikro-orm/core": "5.9.7",
    "@mikro-orm/knex": "5.9.7",
    "@mikro-orm/migrations": "5.9.7",
    "@mikro-orm/postgresql": "5.9.7",
    "@react-email/components": "^0.0.28",
    "@rokmohar/medusa-plugin-meilisearch": "^0.1.1",
    "awilix": "^8.0.1",
    "aws-sdk": "^2.1692.0",
    "bcrypt": "^5.1.1",
    "install": "^0.13.0",
    "medusa-payment-paystack": "^2.0.0",
    "npm": "^11.0.0",
    "pg": "^8.13.0",
    "react-table": "^7.8.0",
    "resend": "^4.0.1",
    "sqs-consumer": "^11.3.0",
    "uuid": "^11.0.3"
  },
  "devDependencies": {
    "@medusajs/test-utils": "latest",
    "@mikro-orm/cli": "5.9.7",
    "@swc/core": "1.5.7",
    "@swc/jest": "^0.2.36",
    "@types/cookie-parser": "^1.4.8",
    "@types/jest": "^29.5.14",
    "@types/node": "^20.0.0",
    "@types/react": "^18.3.2",
    "@types/react-dom": "^18.2.25",
    "jest": "^29.7.0",
    "prop-types": "^15.8.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "ts-jest": "^29.2.5",
    "ts-node": "^10.9.2",
    "typescript": "^5.6.2",
    "vite": "^5.2.11"
  },
  "engines": {
    "node": ">=20"
  }
}
```

### Node.js version

v20.11.0

### Database and its version

postgres:13.5

### Operating system name and version

MacBook Pro , Sonoma 15.2

### Browser name

Chrome

### What happended?

I know in medusa 2, Modules are isolated. But still, I should be able to resolve a module in a function.
I was unable to access the Modules.PAYMENT module inside a function i created. 

 `const paymentModuleService: IPaymentModuleService = container.resolve(Modules.PAYMENT);`
 
This function was added to my custom payment module loader

I was able to access
`const logger = container.resolve(ContainerRegistrationKeys.LOGGER);`

So basically this is what happens. I created the sqsConsumerLoader function and added it as a loader in my custome payment module. This was successful as i was able to run the application without any error.

<img width="735" alt="image" src="https://github.com/user-attachments/assets/0bfdd349-ac7b-4476-abde-4e03d8df4377" />

The idea of the loader is to receive a message from an AWS queue via SQS and run my code when a message is recieved from the AWS queue. So when i send a message to the AWS queue, my application recieves the message then, my application brakes afterword when it gets to resolving Modules.PAYMENT

<img width="809" alt="image" src="https://github.com/user-attachments/assets/9ea06585-bc31-4a48-b2fb-ba08a55b9207" />

and I get this error

<img width="690" alt="image" src="https://github.com/user-attachments/assets/983bcce5-671b-4f15-885f-cde12bb47c9c" />

<img width="786" alt="image" src="https://github.com/user-attachments/assets/f2a27d6e-1d2c-4417-ac11-5578a6bace97" />

@shahednasser
So all i need is how i can resolve the Modules.PAYMENT inside my code

### Expected behavior

I should be able to resolve the Modules.PAYMENT inside my code like i resolved  ContainerRegistrationKeys.LOGGER without my application braking

### Actual behavior

Application brakes when i try to resolve Payment module

<img width="724" alt="image" src="https://github.com/user-attachments/assets/7a5945bc-9714-441c-814a-e772adf9bb40" />

and I get this error

<img width="690" alt="image" src="https://github.com/user-attachments/assets/7609f260-b8e3-4aee-b61a-f2b41f6b4c48" />

### Link to reproduction repo

https://github.com/AsharamiInnovations/boost-backend-v2/tree/eze/ft/sqs

**Comment by helloquang** [comment]
Module is isolated so you can't get other services in your module

In your module you can get your serivice, logger, entity manager, repository, configuration
Reference: https://docs.medusajs.com/resources/medusa-container-resources#module-container-resources

**Comment by shahednasser** [comment]
Hello, as you and the comment above mentioned, you can't resolve other modules in your module due to its isolation.

Can you clarify why you're trying to resolve the payment module in your custom module? Maybe i can help you with finding the appropriate solution.

**Comment by okechukwu0127** [comment]
> Hello, as you and the comment above mentioned, you can't resolve other modules in your module due to its isolation.
> 
> Can you clarify why you're trying to resolve the payment module in your custom module? Maybe i can help you with finding the appropriate solution.

Thanks for your question.
So based on my program logic, 

- Process the webhook data i just recieved by passing it to `paymentModuleService.getWebhookActionAndData` method
- if  result.action  is **captured**, i would trigger `paymentModuleService.capturePayment(captureData)` method
- if result.action is **authorized**, i would trigger  `paymentModuleService.authorizePaymentSession` method

So in a nutshell, I want to leverage the properties of paymentModuleService to complete my code process in my payment circle.

Would appreciate any other workaround on how I can access this methods in the payment Module Service

**Comment by shahednasser** [comment]
As mentioned in [this guide](https://docs.medusajs.com/resources/references/payment/provider#getwebhookactionanddata), you can return an `action` property that tells the payment module what to do. So, if you want to capture the payment, you set `action` to `capture`, and similarly for authorization.

I'll close this issue, feel free to ask more questions if necessary.

**Comment by okechukwu0127** [comment]
> As mentioned in [this guide](https://docs.medusajs.com/resources/references/payment/provider#getwebhookactionanddata), you can return an `action` property that tells the payment module what to do. So, if you want to capture the payment, you set `action` to `capture`, and similarly for authorization.
> 
> I'll close this issue, feel free to ask more questions if necessary.

You did not answer my question on a workaround like I was expecting.

I already know that using paymentModuleService.getWebhookActionAndData, I can get the right actions from my webhook. 

I am expecting you to tell or show me how I can access these methods in payment service maybe via entity manager or any other possible way

**Comment by shahednasser** [comment]
I'm not sure why you need the workaround, can you explain why you need to call these methods from within your module? It's already covered by the payment module so it's not necessary for you to handle it.

**Comment by okechukwu0127** [comment]
> I'm not sure why you need the workaround, can you explain why you need to call these methods from within your module? It's already covered by the payment module so it's not necessary for you to handle it.

When messages are sent from the webhook, i want to handle that flow on my own using using SQS
<img width="805" alt="image" src="https://github.com/user-attachments/assets/f91b4783-d4d1-4303-a388-a7173887468d" />


Can you show me how to access

- getWebhookActionAndData
- capturePayment
- authorizePaymentSession

from the Loader Function `sqsConsumerLoader`

when i do `paymentModuleService.getWebhookActionAndData` the application brakes
when I do `this.getWebhookActionAndData` the application brakes as well

thats the missing part for me

**Comment by shahednasser** [comment]
I would suggest instead to emit the `payment.webhook_received` event in `handleMessage`, as explained [here](https://docs.medusajs.com/learn/fundamentals/events-and-subscribers/emit-event#emit-event-in-a-service) (the docs cover how to resolve the event bus in a service, but you can also resolve it in a loader). Medusa has a subscriber for that event that runs the `getWebhookActionAndData` method then a workflow that captures or authorizes the payment based on the returned action. You can emit the event like this:

```typescript
await eventBus.emit(
      {
        name: "payment.webhook_received",
        data: {
          // this data is passed to `getWebhookActionAndData`
          provider,
          payload: { data: req.body, rawData: req.rawBody, headers: req.headers },
        },
      },
    )
```

Let me know if this makes sense and fits your use case.

**Comment by okechukwu0127** [comment]
> I would suggest instead to emit the `payment.webhook_received` event in `handleMessage`, as explained [here](https://docs.medusajs.com/learn/fundamentals/events-and-subscribers/emit-event#emit-event-in-a-service) (the docs cover how to resolve the event bus in a service, but you can also resolve it in a loader). Medusa has a subscriber for that event that runs the `getWebhookActionAndData` method then a workflow that captures or authorizes the payment based on the returned action. You can emit the event like this:
> 
> ```ts
> await eventBus.emit(
>       {
>         name: "payment.webhook_received",
>         data: {
>           // this data is passed to `getWebhookActionAndData`
>           provider,
>           payload: { data: req.body, rawData: req.rawBody, headers: req.headers },
>         },
>       },
>     )
> ```
> 
> Let me know if this makes sense and fits your use case.

Based on my own program logic, I need to get the wehbook event from SQS handleMessage as I am working. I appreciate your suggestion but again when I tried introducing EVENT Module, i get the same error relating to module isolation in medusa 2. 

**Good News**
I was able to solve the problem by introducing a workflow and step to my application like this.

Inside the workflow steps, i was able to access the PaymentServiceModule without any issues

**SQS handleMessage**
<img width="821" alt="image" src="https://github.com/user-attachments/assets/59619679-55e4-4ffa-baab-660f90e9d26f" />


**THE WORKFLOW**
<img width="639" alt="image" src="https://github.com/user-attachments/assets/d86a2e84-2693-4ece-9b5d-75e5106774f4" />


**THE STEP**
<img width="795" alt="image" src="https://github.com/user-attachments/assets/57d2e685-6612-444f-8ad1-f2bf68224e40" />


I belive this message would be useful to someone else who want to access Module from another module.
@shahednasser Thank you for your time and assistance.

**Comment by shahednasser** [comment]
You shouldn't run a workflow from a loader. Why weren't you able to resolve the event module? Did you follow the last step mentioned [here](https://docs.medusajs.com/learn/fundamentals/events-and-subscribers/emit-event#emit-event-in-a-service)?

**Comment by okechukwu0127** [comment]
@sradevski  @sidd-kulk  @helloquang @shahednasser

I am tying to change the default Order Id in medusa 2.3 to follow a particular format and not sure how to go about that.

Would appreciate if someone can show me how to over right the already existing create order  method in the OrderModuleService such that I can append my  formatted order id to it before an order is then created

I have search for this in the Discord community and nobody seems to have done it. The proposed solution by kapa.ai dont also work.

Would appreciate you assist me with a direction on how to do this
